cEP-0018.md

```diff
@@ -0,0 +1,266 @@
+   generated parser on the fly.
+4. For adding support for a new language, a new Walker class deriving ASTWalker
+   would be added and the rest will be automatically taken care by the library.
+
+## Code Samples/Prototypes
+
+Here is a prototype for the implementations within coantlib:
+
+```python
+from abc import ABC, abstractmethod
+import antlr4
+import coantparsers
+import inspector
+
+from coalib.bearlib.languages.definitions import *
+from coalib.bearlib.languages.Language import parse_lang_str, Languages
+
+
+def resolve_with_package_info(lang):
+    # Use the inspector methods to search amongst all
+    # classes that inherit ASTWalker within this module
+    # and return that class whose supported language matches
+
+class ASTLoader():
+    mapping = {
+        C : [coantparsers.clexer, coantparsers.cparser],
+        Python : [coantparsers.pylexer, coantparsers.pyparser],
+        JavaScript : [coantparsers.jslexer, coantparsers.jsparser],
+        ...
+    }
+
+    @staticmethod
+    def loadFile(file,filename,lang = 'auto'):
+        """
+        Loads file's AST into memory
+        """
+        if(lang == 'auto'):
+          ext = get_extension(filename)
+        else:
+          ext = Languages(parse_lang_str(lang)[0])[0]
+        if ext is None:
+            raise FileExtensionNotSupported('Unknown File Type')
+        inputStream = file.readlines()
+        lexer = mapping[ext][0](inputStream)
+        parser = mapping[ext][1](lexer)
+        return parser
+
+class ASTWalker():
+    treeRoot = None
+    tree = None
+
+    @staticmethod
+    def make_walker():
+        """
+        Resolves walker object for appropriate type
+        """
+        if lang == 'auto':
+            return ASTWalker
+        else:
+            return resolve_with_package_info(lang)
+
+    @staticmethod
+    def get_walker(file, filename, lang):
+        """
+        Returns a walker to the required language
+        """
+        ret_class = make_walker(lang)
+        if not ret_class:
+            raise Exception('Un-supported Language')
+        else:
+            return ret_class(file, filename)
+
+    def __init__(file,filename):
+        self.tree = ASTLoader.load(file, filename)
+        self.treeRoot = self.tree
+
+    def get_next_node(self):
+        """
+        Modify the internal tree variable and return the next node
+        """
+
+    def get_prev_node(self):
+        """
+        Modify the internal tree variable and return prev node
+        """
+
+    def reset(self):
+        """
+        Reset the internal tree variable to tree root
+        """
+
+    def is_terminal_node(node):
+        """
+        Determine if the node is terminal
+        """
+        return isinstance(node, antlr4.tree.Tree.TerminalNode)
+    ...
+```
+
+### Prototype of Py3ASTWalker
+
+These kinds of Walkers will supply a high level walker api
+
+```python
+class BasePyASTWalker(ASTBear):
```

Comment:
ASTWalkers are not Bears as I explained in my earlier comment. A BasePyASTWalker should inherit from ASTWalker and add methods on top of it.

Reply:
Oh
I missed this point
:sweat_smile: 
Will update asap.

Reply:
Updated :sweat_smile: